crates/consensus/consensus/src/lib.rs

```diff
@@ -44,11 +44,11 @@ impl<'a> PostExecutionInput<'a> {
 
 /// Consensus is a protocol that chooses canonical chain.
 #[auto_impl::auto_impl(&, Arc)]
-pub trait Consensus: Debug + Send + Sync {
+pub trait Consensus<H = Header, B = BlockBody>: Debug + Send + Sync {
```

Comment:
imo ideally this has a single block AT bounded by https://github.com/paradigmxyz/reth/blob/97736353447e4e5a9f9323ecf3b5839fe7ef8190/crates/primitives-traits/src/block/mod.rs#L22

`Consensus` responsibility is to validate blocks so don't think it should know about entire node storage model. this would also be consistent with pool traits which only know the transaction AT

I've done this as 2 generics right now so that we can still keep `dyn Consensus` everywhere

does it make sense?

Reply:
why can't you keep the trait object if you add the generic?

`NetworkPrimitives` has header and body as ATs, and the block, so I'm not sure you're intentions are really clear?

Reply:
> why can't you keep the trait object if you add the generic?

I mean that we can't keep it as is if we'd add AT because we'd need to specify `Block = Block` everywhere, so went with generics for now

> `NetworkPrimitives` has header and body as ATs, and the block, so I'm not sure you're intentions are really clear?

sorry, not sure I follow, how are `NetworkPrimitives` related to consensus trait?

Reply:
yeah a single AT would be ideal, but for now we can introduce both AT just to move forward

Reply:
i

> > why can't you keep the trait object if you add the generic?
> 
> I mean that we can't keep it as is if we'd add AT because we'd need to specify `Block = Block` everywhere, so went with generics for now
> 

I see, seems to be a misunderstanding, I didn't mean adding an AT to the consensus trait

> > `NetworkPrimitives` has header and body as ATs, and the block, so I'm not sure your intentions are really clear?
> 
> sorry, not sure I follow, how are `NetworkPrimitives` related to consensus trait?

if you remove the rlp encoding trait bounds from ATs on `NetworkPrimitives`, and just have auto trait bounds, then the associated types can easily just be the unit struct `()`.

by passing one type, that aggregates the data primitives as associated types, we make a framework that is extensible without breaking the public api.

in the future, and even towards the end of this sdk integration task, it may well be, that somewhere in consensus, another data primitive type is needed. then we have to change the type definition again to put another generic to the list if we do it the way you impl it now.

Reply:
just recently op impl of transaction rpc, needed a receipt type, who would have known.

https://github.com/paradigmxyz/reth/issues/11164
https://github.com/paradigmxyz/reth/pull/11732

keep the code usable for use cases you could not have had time to think of yourself at the time of impl.